File: javascript/arrays/two_sum.js
var twoSum = function (nums, target) {
  const indexMap = {};

  for (let i = 0; i < nums.length; i++) {
    const currentIndexVal = indexMap[nums[i]];

    // If we have the other pair in our map then we found the target
    if (currentIndexVal >= 0) return [currentIndexVal, i];
    else {
      // Storing the num we want to find on our map
      const numberToFind = target - nums[i];
      indexMap[numberToFind] = i;
    }
  }

  return null;
};

File: arrays/prefix_sum/two_sum.py
class Solution:
    def twoSum(self, nums: List[int], target: int) -> List[int]:
        prev_map = {} # val : index

        for i, num in enumerate(nums):
            curr_diff = target - num

            # If we see the value again, we have found the two sum
            if curr_diff in prev_map:
                return [prev_map[curr_diff], i]

            # Recording the value and index
            prev_map[num] = i

        return # Not really necessary since we're guaranteed a sol.

# Time Complexity: O(n)
# Space Complexity: O(n)
